File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.py
def solution(land):
    dp = [[0 for _ in range(4)] for _ in range(len(land))]
    
    for col in range(len(land)):
        for row in range(4):
            if col == 0:
                dp[0][row] = land[0][row]
            else:
                # print(col, row)
                for i in range(4):
                    if i != row:
                        dp[col][row] = max(dp[col][row], land[col][row] + dp[col-1][i])
                        
    answer = 0
    print(dp[-1])
    
    for i in dp[-1]:
        answer = max(answer, i)
        
    return answer

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.js
function solution(land) {
    const dp = Array.from(Array(land.length), () => Array(land[0].length).fill(0))
    
    for (let col = 0; col < 4; col++) {
        dp[0][col] = land[0][col]
    }
    
    for (let row = 1; row < land.length; row++) {
        for (let col = 0; col < 4; col++) {
               // col = 0 -> 1, 2, 3 => 1, 2, 3
                // 1 -> 0, 2, 3... => (2, 3, 4) % 4
                // 2 -> 0, 1, 3 => (3, 4, 5) % 4
            let plusNum = 0
            let temp = col
            while (plusNum < 3) {
                temp = (temp + 1) % 4
                dp[row][col] = Math.max(dp[row][col], dp[row - 1][temp])
                plusNum += 1
            }
            dp[row][col] += land[row][col]
        }
    }
    
    return Math.max(...dp[land.length - 1])
}